### PR TITLE
(Probably ready for test merge) DYNAMIC 2021

### PR DESCRIPTION
**Pull request by keronshb**
## About The Pull Request

This is Dynamic from TG being ported here.  Hopefully this works.  It needs live testing.

It's a myriad of PRs so I've even considered calling it a hardsync.  

Also adds Mob Lists since dynamic needed it to work.

Keeps Dynamic/Extended voting.

## Why It's Good For The Game

Everyone's been asking for dynamic so here it is.  This system is objectively better instead of the old secret/extended system.

## Changelog
:cl:
add: Revs winning/losing doesn't end the mode
code: Dynamic replaces secret
config: added configs related to dynamic, removed configs from old dynamic
/:cl:

### CODE REVIEW

File: code/game/gamemodes/dynamic/dynamic_rulesets_latejoin.dm [hunk 1]
```diff
@@ -47,8 +41,6 @@
 	assigned += M.mind
 	M.mind.special_role = antag_flag
 	M.mind.add_antag_datum(antag_datum)
-	log_admin("[M.name] was made into a [name] by dynamic.")
```
Comment (Putnam3145) [inline code/game/gamemodes/dynamic/dynamic_rulesets_latejoin.dm:50]: probably shouldn't actually remove this

File: code/game/gamemodes/dynamic/dynamic_rulesets_latejoin.dm [hunk 2]
```diff
@@ -125,163 +113,45 @@
...
-				var/datum/antagonist/rev/head/R = M.has_antag_datum(/datum/antagonist/rev/head)
-				R.remove_revolutionary(FALSE, "gamemode")
-				var/mob/living/carbon/C = M.current
-				if(C.stat == DEAD)
-					C.makeUncloneable()
-			if(M.has_antag_datum(/datum/antagonist/rev))
-				var/datum/antagonist/rev/R = M.has_antag_datum(/datum/antagonist/rev)
-				R.remove_revolutionary(FALSE, "gamemode")
-		return RULESET_STOP_PROCESSING
+	var/winner = revolution.process_victory(revs_win_threat_injection)
```
Comment (Putnam3145) [inline code/game/gamemodes/dynamic/dynamic_rulesets_latejoin.dm:129]: probably shouldn't change this at all unless you want to port all of the revs changes tg's done too

File: code/game/gamemodes/dynamic/dynamic.dm [hunk 3]
```diff
@@ -755,63 +676,82 @@ GLOBAL_VAR_INIT(dynamic_forced_storyteller, null)
...
 		for(var/blocking in blocking_list)
-			for(var/datum/executed in rule_list)
+			for(var/_executed in rule_list)
+				var/datum/executed = _executed
 				if(blocking == executed.type)
 					return TRUE
 	return FALSE
 
-/// Checks if client age is age or older.
-/datum/game_mode/dynamic/proc/check_age(client/C, age)
```
Comment (Putnam3145) [inline code/game/gamemodes/dynamic/dynamic.dm:764]: you actually need this, not sure why it's being removed

File: code/game/gamemodes/dynamic/dynamic.dm [hunk 4]
```diff
@@ -188,434 +190,363 @@ GLOBAL_VAR_INIT(dynamic_forced_storyteller, null)
...
 		for(var/datum/station_goal/G in station_goals)
 			G.on_report()
 			. += G.get_report()
 
-	print_command_report(., "Central Command Status Summary", announce=FALSE)
-	if(GLOB.security_level >= SEC_LEVEL_BLUE)
-		priority_announce("A summary has been copied and printed to all communications consoles.", "Security level elevated.", "intercept")
-	else
-		priority_announce("Thanks to the tireless efforts of our security and intelligence divisions, there are currently no likely threats to [station_name()]. All station construction projects have been authorized. Have a secure shift!", "Security Report", "commandreport")
+	if(SSstation.station_traits.len)
```
Comment (Putnam3145) [inline code/game/gamemodes/dynamic/dynamic.dm]: we don't have station traits

Comment (keronshb) [reply code/game/gamemodes/dynamic/dynamic.dm]: Removed.

File: code/game/gamemodes/dynamic/dynamic.dm (hunk 3 above)
Comment (keronshb) [reply code/game/gamemodes/dynamic/dynamic.dm:764]: I tried re adding this but it did nothing.  I'm still getting errors.  Tried adding it in some other folder, gave even more errors.

File: code/game/gamemodes/dynamic/dynamic_rulesets_latejoin.dm (hunk 1 above)
Comment (keronshb) [reply code/game/gamemodes/dynamic/dynamic_rulesets_latejoin.dm:50]: readded

File: config/dynamic.json [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+
+		"Nuclear Emergency": {
+			"weight": 0
+		},
+
+		"Revolution": {
+			"weight": 0
+		},
+
+		"Families": {
```
Comment (zeroisthebiggay) [inline config/dynamic.json]: we don't have families??????????

Comment (keronshb) [reply config/dynamic.json]: We don't.

I'll add it in at some point but I need to see if dynamic works.

**Comment by buffyuwu** [comment]
oh word someone should port families

File: code/game/gamemodes/dynamic/dynamic_rulesets_midround.dm [hunk 6]
```diff
@@ -133,25 +104,24 @@
 	message_admins("Polling [possible_volunteers.len] players to apply for the [name] ruleset.")
 	log_game("DYNAMIC: Polling [possible_volunteers.len] players to apply for the [name] ruleset.")
 
-	candidates = pollGhostCandidates("The mode is looking for volunteers to become a [name]", antag_flag, SSticker.mode, antag_flag, poll_time = 300)
+	candidates = pollGhostCandidates("The mode is looking for volunteers to become [antag_flag] for [name]", antag_flag, antag_flag_override ? antag_flag_override : antag_flag, poll_time = 300)
```
Comment (Putnam3145) [inline code/game/gamemodes/dynamic/dynamic_rulesets_midround.dm]:
```suggestion
	candidates = pollGhostCandidates("The mode is looking for volunteers to become [antag_flag] for [name]", antag_flag, be_special_flag = antag_flag_override ? antag_flag_override : antag_flag, poll_time = 300)
```

File: code/game/gamemodes/dynamic/dynamic.dm [hunk 7]
```diff
@@ -282,340 +285,264 @@ GLOBAL_VAR_INIT(dynamic_forced_storyteller, null)
...
 /datum/game_mode/dynamic/proc/generate_threat()
-	var/relative_threat = LORENTZ_DISTRIBUTION(GLOB.dynamic_curve_centre, GLOB.dynamic_curve_width)
-	threat_level = round(lorentz_to_threat(relative_threat), 0.1)
-
-	peaceful_percentage = round(LORENTZ_CUMULATIVE_DISTRIBUTION(relative_threat, GLOB.dynamic_curve_centre, GLOB.dynamic_curve_width), 0.01)*100
-
-	SSblackbox.record_feedback("tally","dynamic_threat",threat_level,"Initial threat level")
-	SSblackbox.record_feedback("tally","dynamic_threat",GLOB.dynamic_curve_centre,"Curve centre")
-	SSblackbox.record_feedback("tally","dynamic_threat",GLOB.dynamic_curve_width,"Curve width")
-	SSblackbox.record_feedback("tally","dynamic_threat",peaceful_percentage,"Percent of same-vote rounds that are more peaceful")
```
Comment (Putnam3145) [inline code/game/gamemodes/dynamic/dynamic.dm:340]: add these back, after peaceful_percentage is assigned, though change "same-vote" to "same-center"

File: code/game/gamemodes/dynamic/dynamic.dm [hunk 8]
```diff
@@ -282,340 +285,264 @@ GLOBAL_VAR_INIT(dynamic_forced_storyteller, null)
...
+/datum/game_mode/dynamic/proc/setup_parameters()
 	log_game("DYNAMIC: Dynamic mode parameters for the round:")
-	log_game("DYNAMIC: Centre is [GLOB.dynamic_curve_centre], Width is [GLOB.dynamic_curve_width], Forced extended is [GLOB.dynamic_forced_extended ? "Enabled" : "Disabled"], No stacking is [GLOB.dynamic_no_stacking ? "Enabled" : "Disabled"].")
-	log_game("DYNAMIC: Stacking limit is [GLOB.dynamic_stacking_limit], Classic secret is [GLOB.dynamic_classic_secret ? "Enabled" : "Disabled"], High population limit is [GLOB.dynamic_high_pop_limit].")
+	log_game("DYNAMIC: Centre is [threat_curve_centre], Width is [threat_curve_width], Forced extended is [GLOB.dynamic_forced_extended ? "Enabled" : "Disabled"], No stacking is [GLOB.dynamic_no_stacking ? "Enabled" : "Disabled"].")
+	log_game("DYNAMIC: Stacking limit is [GLOB.dynamic_stacking_limit].")
 	if(GLOB.dynamic_forced_threat_level >= 0)
 		threat_level = round(GLOB.dynamic_forced_threat_level, 0.1)
-		threat = threat_level
-		SSblackbox.record_feedback("tally","dynamic_threat",threat_level,"Threat level (forced)")
```
Comment (Putnam3145) [inline code/game/gamemodes/dynamic/dynamic.dm:358]: add this one back

File: code/game/gamemodes/dynamic/dynamic.dm [hunk 9]
```diff
@@ -282,340 +285,264 @@ GLOBAL_VAR_INIT(dynamic_forced_storyteller, null)
...
 			candidates.Add(player)
 	log_game("DYNAMIC: Listing [roundstart_rules.len] round start rulesets, and [candidates.len] players ready.")
 	if (candidates.len <= 0)
 		log_game("DYNAMIC: [candidates.len] candidates.")
 		return TRUE
-	if (roundstart_rules.len <= 0)
-		log_game("DYNAMIC: [roundstart_rules.len] rules.")
-		return TRUE
-	SSblackbox.record_feedback("tally","dynamic",roundstart_rules.len,"Roundstart rules considered")
-	SSblackbox.record_feedback("tally","dynamic",roundstart_pop_ready,"Players readied up")
```
Comment (Putnam3145) [inline code/game/gamemodes/dynamic/dynamic.dm:395]: these two should probably be there

File: code/game/gamemodes/dynamic/dynamic.dm [hunk 10]
```diff
@@ -282,340 +285,264 @@ GLOBAL_VAR_INIT(dynamic_forced_storyteller, null)
...
-	return FALSE
+		stack_trace("The starting rule \"[ruleset.name]\" failed to pre_execute.")
+	return 0
 
 /// Mainly here to facilitate delayed rulesets. All roundstart rulesets are executed with a timered callback to this proc.
 /datum/game_mode/dynamic/proc/execute_roundstart_rule(sent_rule)
 	var/datum/dynamic_ruleset/rule = sent_rule
 	if(rule.execute())
-		current_rules += rule
-		SSblackbox.record_feedback("associative","dynamic_rulesets",1,rule.get_blackbox_info())
```
Comment (Putnam3145) [inline code/game/gamemodes/dynamic/dynamic.dm:573]: technically this one should be there but if you removed rule.get_blackbox_info() don't bother

File: code/game/gamemodes/dynamic/dynamic.dm [hunk 11]
```diff
@@ -630,119 +557,112 @@ GLOBAL_VAR_INIT(dynamic_forced_storyteller, null)
...
+			new_rule.pre_execute(population)
 			if (new_rule.execute()) // This should never fail since ready() returned 1
-				if(new_rule.flags & HIGHLANDER_RULESET)
-					highlander_executed = TRUE
+				if(new_rule.flags & HIGH_IMPACT_RULESET)
+					high_impact_ruleset_executed = TRUE
 				else if(new_rule.flags & ONLY_RULESET)
 					only_ruleset_executed = TRUE
 				log_game("DYNAMIC: Making a call to a specific ruleset...[new_rule.name]!")
-				SSblackbox.record_feedback("associative","dynamic_rulesets",1,new_rule.get_blackbox_info())
```
Comment (Putnam3145) [inline code/game/gamemodes/dynamic/dynamic.dm:650]: same, if you removed get_blackbox_info don't bother

File: code/game/gamemodes/dynamic/dynamic.dm [hunk 12]
```diff
@@ -630,119 +557,112 @@ GLOBAL_VAR_INIT(dynamic_forced_storyteller, null)
...
 /datum/game_mode/dynamic/process()
-	if (pop_last_updated < world.time - (120 SECONDS))
-		pop_last_updated = world.time
-		update_playercounts()
-
 	for (var/datum/dynamic_ruleset/rule in current_rules)
 		if(rule.rule_process() == RULESET_STOP_PROCESSING) // If rule_process() returns 1 (RULESET_STOP_PROCESSING), stop processing.
 			current_rules -= rule
-			SSblackbox.record_feedback("tally","dynamic",1,"Rulesets finished")
-			SSblackbox.record_feedback("associative","dynamic_rulesets_finished",1,rule.get_blackbox_info())
```
Comment (Putnam3145) [inline code/game/gamemodes/dynamic/dynamic.dm:691]: for sure add the first one, but ditto with get_blackbox_info

File: code/game/gamemodes/dynamic/dynamic_rulesets_midround.dm (hunk 6 above)
Comment (keronshb) [reply code/game/gamemodes/dynamic/dynamic_rulesets_midround.dm]: Fixed

File: code/game/gamemodes/dynamic/dynamic.dm (hunk 10 above)
Comment (keronshb) [reply code/game/gamemodes/dynamic/dynamic.dm:573]: It's gone

File: code/game/gamemodes/dynamic/dynamic.dm (hunk 11 above)
Comment (keronshb) [reply code/game/gamemodes/dynamic/dynamic.dm:650]: Also gone

File: code/game/gamemodes/dynamic/dynamic.dm (hunk 12 above)
Comment (keronshb) [reply code/game/gamemodes/dynamic/dynamic.dm:691]: Added the first, kept second removed

File: code/game/gamemodes/dynamic/dynamic.dm (hunk 9 above)
Comment (keronshb) [reply code/game/gamemodes/dynamic/dynamic.dm:395]: Added

File: code/game/gamemodes/dynamic/dynamic.dm (hunk 7 above)
Comment (keronshb) [reply code/game/gamemodes/dynamic/dynamic.dm:340]: Added and tweaked

File: code/game/gamemodes/dynamic/dynamic_rulesets_midround.dm [hunk 13]
```diff
@@ -315,21 +300,20 @@
...
 	required_candidates = 1
-	weight = 1
+	weight = 2
 	cost = 20
-	requirements = list(90,90,70,50,50,50,50,40,30,30)
-	high_population_requirement = 30
+	requirements = list(101,101,100,80,50,30,20,10,10,10)
 	repeatable = TRUE
-	property_weights = list("story_potential" = 2, "trust" = 1, "chaos" = 2, "extended" = -2)
-	var/datum/mind/wizard
```
Comment (Putnam3145) [inline code/game/gamemodes/dynamic/dynamic_rulesets_midround.dm:330]: put this line back in

File: code/game/gamemodes/dynamic/dynamic_rulesets_midround.dm [hunk 14]
```diff
@@ -339,20 +323,6 @@
 /datum/dynamic_ruleset/midround/from_ghosts/wizard/finish_setup(mob/new_character, index)
 	..()
 	new_character.forceMove(pick(GLOB.wizardstart))
-	wizard = new_character.mind
```
Comment (Putnam3145) [inline code/game/gamemodes/dynamic/dynamic_rulesets_midround.dm:342]: this one too

File: code/game/gamemodes/dynamic/dynamic_rulesets_midround.dm [hunk 15]
```diff
@@ -339,20 +323,6 @@
...
-
-/datum/dynamic_ruleset/midround/from_ghosts/wizard/rule_process() // i can literally copy this from are_special_antags_dead it's great
```
Comment (Putnam3145) [inline code/game/gamemodes/dynamic/dynamic_rulesets_midround.dm:344]: keep this whole proc

**Pull request merged**